# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.drop('spam', axis=1)


In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

In [12]:
# Split the data into X_train, X_test, y_train, y_test using a 80-20 split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)
#show the first 5 rows of the training set
df.head(10)


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1947            0.00               0.00           0.00           0.0   
2159            0.08               0.17           0.17           0.0   
4223            0.00               0.00           1.85           0.0   
2624            0.00               0.00           0.00           0.0   
2743            0.00               0.00           0.00           0.0   
3451            0.21               0.00           0.21           0.0   
464             0.00               0.00           0.00           0.0   
3219            0.00               0.00           0.87           0.0   
3007            0.05               0.00           0.15           0.0   
1549            0.22               0.00           0.78           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1947            0.0            0.00              0.00                0.00   
2159            0.8            0.08              0.00                0.26   
4223            0.0            0.00              0.00                0.00   
2624            0.0            0.00              0.00                0.00   
2743            0.0            0.00              0.00                0.39   
3451            0.0            0.00              0.00                0.00   
464             0.0            0.00              0.00                0.00   
3219            0.0            2.63              0.00                0.00   
3007            0.1            0.00              0.00                0.00   
1549            0.0            0.11              0.11                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1947             0.00            0.00  ...                  0.00        0.000   
2159             0.00            0.00  ...                  0.00        0.023   
4223             0.00            1.85  ...                  0.00        0.000   
2624             0.00            0.00  ...                  0.00        0.000   
2743             0.00            0.00  ...                  0.00        0.000   
3451             0.00            0.00  ...                  0.00        0.000   
464              0.00            0.00  ...                  0.00        0.000   
3219             0.00            0.00  ...                  0.00        0.317   
3007             0.52            0.00  ...                  0.05        0.164   
1549             0.22            0.11  ...                  0.00        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1947        0.000        0.000        0.000        0.000        0.380   
2159        0.046        0.000        0.000        0.023        0.000   
4223        0.000        0.000        0.692        0.000        0.000   
2624        0.000        0.000        0.000        0.000        0.000   
2743        0.049        0.000        0.000        0.000        0.000   
3451        0.084        0.000        0.028        0.084        0.084   
464         0.000        0.000        0.486        0.000        0.000   
3219        0.317        0.000        0.000        0.000        0.000   
3007        0.171        0.068        0.000        0.013        0.000   
1549        0.142        0.000        0.775        0.224        0.142   

      capital_run_length_average  capital_run_length_longest  \
1947                       2.307                           9   
2159                       2.658                          57   
4223                       1.727                           5   
2624                       1.000                           1   
2743                       2.000                          38   
3451                       4.110                          62   
464                        2.681                          11   
3219                       1.269                           5   
3007                       3.591                          35   
1549                       5.

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [13]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()


In [14]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [15]:
# Scale the training data
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns, index=X_train.index)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns, index=X_test.index)

# Display the first few rows of the scaled training data
print("Scaled training data:")
print(X_train_scaled.head())


Scaled training data:
      word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1947       -0.358493          -0.166316      -0.561452     -0.046223   
2159       -0.086769          -0.034973      -0.223847     -0.046223   
4223       -0.358493          -0.166316       3.112489     -0.046223   
2624       -0.358493          -0.166316      -0.561452     -0.046223   
2743       -0.358493          -0.166316      -0.561452     -0.046223   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1947      -0.462132       -0.348265         -0.289743           -0.254338   
2159       0.742377       -0.062859         -0.289743            0.380197   
4223      -0.462132       -0.348265         -0.289743           -0.254338   
2624      -0.462132       -0.348265         -0.289743           -0.254338   
2743      -0.462132       -0.348265         -0.289743            0.697464   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [17]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(random_state=1)
logistic_model.fit(X_train_scaled, y_train)

# Make predictions
testing_predictions = logistic_model.predict(X_test_scaled)

# Calculate and print accuracy score
accuracy = accuracy_score(y_test, testing_predictions)
print(f"Logistic Regression Model Accuracy: {accuracy:.4f}")



Logistic Regression Model Accuracy: 0.9273


In [18]:
# Make and save testing predictions with the trained Logistic Regression model using the test data
lr_predictions = logistic_model.predict(X_test_scaled)

# Review the predictions
print("First 10 predictions:", lr_predictions[:10])
print("\nActual first 10 values:", y_test[:10].values)
print(f"\nTotal predictions: {len(lr_predictions)}")


First 10 predictions: [0 0 1 0 0 1 1 0 1 1]

Actual first 10 values: [1 1 1 0 0 0 0 0 1 1]

Total predictions: 921


In [19]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, testing_predictions)
print(f"Logistic Regression Model Accuracy: {accuracy:.4f}")


Logistic Regression Model Accuracy: 0.9273


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [21]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(random_state=1)
rf_model.fit(X_train_scaled, y_train)

# Make predictions
rf_predictions = rf_model.predict(X_test_scaled)

# Calculate and print accuracy score
rf_accuracy = accuracy_score(y_test, rf_predictions)
print(f"Random Forest Model Accuracy: {rf_accuracy:.4f}")

Random Forest Model Accuracy: 0.9587


In [22]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data
rf_testing_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
print("First 10 Random Forest predictions:", rf_predictions[:10])
print("\nActual first 10 values:", y_test[:10].values)
print(f"\nTotal Random Forest predictions: {len(rf_predictions)}")


First 10 Random Forest predictions: [1 1 1 0 0 1 0 0 1 1]

Actual first 10 values: [1 1 1 0 0 0 0 0 1 1]

Total Random Forest predictions: 921


In [23]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, testing_predictions)
print(f"Logistic Regression Model Accuracy: {accuracy:.4f}")


Logistic Regression Model Accuracy: 0.9273


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Although both model's predictions showed favorable results, the Random Forest Classifier model yielded a higher accuracy of detecting spam at 95.87% compared to the logistic regression model's 92.73%*